### ISSUE TITLE
Prestressed modal analysis with volume elements

### ISSUE THREAD

**Issue by ste-lor** [KratosMultiphysics/Kratos#9550]
I'm currently doing a small project implementing a pre-stressed modal analysis for a centrifugal compressor using volume elements. Only the centrifugal force is to be considered as a force.
First, the stresses as a function of the rot. speed are calculated, then the eigenfrequencies with the respective preload should be calculated. The calculation of sigma = f(rpm) does work, the second part not. Meaning that the eigenfrequency do not change as function of the rot. speed.

This question is very similar to: #6932
Here they changed the "model_import_settings" of the .json-file (referring to the modal analysis) to:
 "model_import_settings"              : {
                    "input_type"       : "use_input_model_part"
                },
However, they used shell elements. The last comment in #6932 says that this procedure does not work when not using a rotational-like displacement. Is there a method for volume elements?

Attached you find the .py-code and the two .json-files, to better represent what I mean.
[files.zip](https://github.com/KratosMultiphysics/Kratos/files/7888776/files.zip)
.

**Comment by loumalouomega** [comment]
Huuuum, do you mean that the prestress is nor affecting the eigenfrequencies?

**Comment by ste-lor** [comment]
Yes, exactly. I calculate first the stresses at a certain speed, and then I would like to apply on this condition a modal analysis. The eigenfrequency is expected to increase (with respect to the static case rpm=0), but stays constant in my case.
It seems that Kratos does not use the data under prestressed condition but always the static one.

For shell elements, substituting the command "use_import_model_part" for the input_type solves this issue. However, this somehow does not work for volume elements. Any ideas what I could have done wrong?

**Comment by loumalouomega** [comment]
Huuum, something tells me that we are not taking the prestress into consideration when computing the stiffness matrix (should be done via the CL, or internal implementation of the element) @KratosMultiphysics/structural-mechanics , any hint?

**Comment by AlejandroCornejo** [comment]
Dear @ste-lor Can you explain a bit more what do you need, what do you did and the problems encountered? It's difficult for me to understand correctly this issue :)

**Comment by ste-lor** [comment]
I think @loumalouomega is right, the additional stiffness matrix due to the prestress is not considered, but I do not know how to do that.
Yes sure, thanks for your time.
I think the easiest way is to explain it with the written python code - you can find it attached in the first comment.
Starting with line 47: Here, the stresses of the model at a specific rot. speed are calculated (considering centrifugal force). To do that, the .json-file (called "ProjectParametersPrestressed") is read, which itself refers to the file containing the data points of the meshed geometry (in my case called compressor.mdpa; can be seen in line 16 and 17 of the .json-file "ProjectParametersPrestressed").
Next, but always in the same loop (starting at line 62) I would like to do a modal analysis. Therefore, I read an other .json-file (called "ProjectParametersModal") in which the properties of the modal analysis are defined. The thing is now, that I do not want to apply the modal analysis on the initial geometry (by reading the compressor.mdpa file) but on the prestressed one. 

I was thinking, that this could be done by modifying inside the .json-file "ProjectParametersModel" the line 15. Meaning that changing the "input_type" from "mdpa" to "use_input_model_part". This approach does work for shell elements as also described in #6932.
Maybe for volume elements some additional information have to be implemented?

**Comment by e-dub** [comment]
Thanks for the responses!

We would like to prestress the structure and calculate the eigenfrequency of the loaded structure. 

Linearly this would include the geometrical stiffness,
![Figure](https://latex.codecogs.com/png.image?\dpi{110}&space;\bg_white&space;\left(-\omega_{i}^{2}\underline{\underline{m}}+\left(\underline{\underline{k}}_{E}+\underline{\underline{k}}_{G}\right)\right)\underline{\Phi}_{i}=\underline{0}).

With nonlinear prestress analysis, this would take consideration into changes in the mass and stiffness matrices due to the deformation,
![Figure](https://latex.codecogs.com/png.image?\dpi{110}&space;\bg_white&space;\left(-\omega_{i}^{2}\underline{\underline{m}}\left({\underline{u}}\right)+\underline{\underline{k}}\left({\underline{u}}\right)\right)\underline{\Phi}_{i}=\underline{0}).

If I am not mistaken, the geometrical (prestress) stiffness matrix has not been implemented, i.e. for eigenvalue buckling. The nonlinear route though should be possible (I think).  

The idea was to load the structure, calculate nonlinear quasistatic, keep the model and its responses and then carry out the eigenfrequency analysis.  Maybe I am missing something somewhere. 

We are currently using the following:
- SmallDisplacementElement3D4N
- LinearElastic3DLaw

**Comment by loumalouomega** [comment]
> Thanks for the responses!
> 
> We would like to prestress the structure and calculate the eigenfrequency of the loaded structure.
> 
> Linearly this would include the geometrical stiffness matrix, ![Figure](https://camo.githubusercontent.com/61cc7e7cd312190446593273e8407607a78740bf9c8c0922af4c77a005d46f72/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f706e672e696d6167653f2535436470692537423131302537442673706163653b25354362675f77686974652673706163653b2535436c656674282d2535436f6d6567615f2537426925374425354525374232253744253543756e6465726c696e65253742253543756e6465726c696e652537426d2537442537442b2535436c65667428253543756e6465726c696e65253742253543756e6465726c696e652537426b2537442537445f253742452537442b253543756e6465726c696e65253742253543756e6465726c696e652537426b2537442537445f25374247253744253543726967687429253543726967687429253543756e6465726c696e652537422535435068692537445f253742692537443d253543756e6465726c696e6525374230253744).
> 
> With nonlinear prestress analysis, this would take changes in the mass and stiffness matrices due to the deformation into consideration, ![Figure](https://camo.githubusercontent.com/70892f5f8a3c76e04e6743737d88d88b2a7a90e15b0631b126aa7b7bbcac702c/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f706e672e696d6167653f2535436470692537423131302537442673706163653b25354362675f77686974652673706163653b2535436c656674282d2535436f6d6567615f2537426925374425354525374232253744253543756e6465726c696e65253742253543756e6465726c696e652537426d2537442537442535436c65667428253742253543756e6465726c696e65253742752537442537442535437269676874292b253543756e6465726c696e65253742253543756e6465726c696e652537426b2537442537442535436c65667428253742253543756e6465726c696e6525374275253744253744253543726967687429253543726967687429253543756e6465726c696e652537422535435068692537445f253742692537443d253543756e6465726c696e6525374230253744).
> 
> If I am not mistaken, the geometrical (prestress) stiffness matrix has not been implemented, i.e. for eigenvalue buckling. The nonlinear route though should be possible (I think).
> 
> The idea was to load the structure, calculate nonlinear quasistatic, keep the model and its responses and then carry out the eigenfrequency analysis. Maybe I am missing something somewhere.
> 
> We are currently using the following:
> 
>     * SmallDisplacementElement3D4N
> 
>     * LinearElastic3DLaw

Exactly, currently prestress is not implemented. But we can do it with CL with the interface implemented by @AlejandroCornejo , similar to the one used in https://github.com/KratosMultiphysics/Kratos/pull/9053

**Comment by e-dub** [comment]
https://github.com/KratosMultiphysics/Kratos/pull/9053  is a big one... I see the parallels to an initial thermal analysis, but would we have to implement a CL to store the prestress or would this interface do it for us?
